src/contracts/atlas/AtlasVerification.sol

```diff
@@ -548,27 +549,21 @@ contract AtlasVerification is EIP712, DAppIntegration, AtlasConstants {
     function _verifyUser(
         DAppConfig memory dConfig,
         UserOperation calldata userOp,
+        bytes32 userOpHash,
         address msgSender,
         bool isSimulation
     )
         internal
         returns (ValidCallsResult)
     {
+        if (userOp.from == address(this) || userOp.from == ATLAS || userOp.from == userOp.control) {
+            return ValidCallsResult.UserFromInvalid;
+        }
+
         // Verify the signature before storing any data to avoid
         // spoof transactions clogging up dapp userNonces
 
-        bool isFromContract = userOp.from.code.length > 0;
-
-        bool signatureValid = false;
-        if (isFromContract) {
-            if (userOp.from == address(this) || userOp.from == ATLAS || userOp.from == userOp.control) {
-                return ValidCallsResult.UserFromInvalid;
-            }
-            signatureValid = IAccount(userOp.from).validateUserOp(userOp, _getUserOpHash(userOp), 0) == 0;
-        } else {
-            // user is an EOA
-            signatureValid = _verifyUserSignature(userOp);
-        }
+        bool signatureValid = SignatureChecker.isValidSignatureNow(userOp.from, _hashTypedDataV4(_getUserOpHash(userOp)), userOp.signature);
 
         bool userIsBundler = userOp.from == msgSender;
         bool hasNoSignature = userOp.signature.length == 0;
```

Comment:
```suggestion
        bool userIsBundler = userOp.from == msgSender;
        bool hasNoSignature = userOp.signature.length == 0;
        bool signatureValid = hasNoSignature
            ? false
            : SignatureChecker.isValidSignatureNow(userOp.from, _hashTypedDataV4(_getUserOpHash(userOp)), userOp.signature);
```
Could move the sig check below, then only do the check if there is a signature to check, to save gas

Reply:
I don't think a ERC-1271 message has to actually have anything in the message field to be valid, this function just has to return true. So for example, the target contract could have a map of known good hashes rather than signatures, and validate against this. In this case, this protection would break contracts that follow this pattern.

Reply:
True, as per the 1271 spec, the smart wallet can technically tell us an empty string is a valid sig. Disregard this suggestion then